# Laboratorio de regresión logística

|                |   |
:----------------|---|
| **Nombre**     | Jesús Emmanuel Flores Cortés  |
| **Fecha**      | 13/10/2025  |
| **Expediente** | 751571  | 

In machine learning, Support Vector Machines (SVM) are supervised learning models with associated learning algorithms that analyze data used for classification and regression analysis. It is mostly used in classification problems. In this algorithm, each data item is plotted as a point in p-dimensional space (where p is the number of features), with the value of each feature being the value of a particular coordinate. Then, classification is performed by finding the hyper-plane that best differentiates the two classes (or more if we have a multi class problem):

$$ f(x) = w^T \varphi(x) + b $$

where $\varphi: X \rightarrow F $ is a function that makes each input point $x$ correspond to a point in F, where F is a Hilbert space.

In addition to performing linear classification, SVMs can efficiently perform a non-linear classification, implicitly mapping their inputs into high-dimensional feature spaces (more specifically using the kernel trick, like the RBF funcion). 

[1]

OLS utilizes the squared residuals to fit the parameters. Large residuals caused by outliers may worsen the accuracy significantly.

Support Vectors use piecewise linear functions to counter this, in which a hyperparameter  $\epsilon$ called the margin lets errors that are less or equal to it be 0, and error larger than it be $e - \epsilon$. 

The problem to solve is:

\begin{split}
        \min_{w, b, \xi, \xi^*} \mathcal{P}_\epsilon(w, b, \xi) &= \frac{1}{2} w^T w + c \sum_{k=1}^{N} \xi_k \\
        \text{s.t. } & y_k [w^T \varphi(x_k) - b] \geq 1- \xi_k,\ \ k = 1, ..., N \\
        & \xi_k \geq 0,\ \ k = 1, ..., N
\end{split}


The most important question that arises when using a SVM is how to choose the correct hyperplane. Consider the following scenarios:

### Scenario 1

In this scenario there are three hyperplanes called A, B, and C. Now, the problem is to identify the hyperplane which best differentiates the stars and the circles.

<center><img src="https://media.geeksforgeeks.org/wp-content/uploads/SVM_21-2.png" alt="what image shows"></center>

In this case, hyperplane B separates the stars and the circle betters, hence it is the correct hyperplane.


### Scenario 2

Now take another scenario where all three hyperplanes are segregating classes well. The question that arises is how to choose the best hyperplane in this situation.

<center><img src="https://media.geeksforgeeks.org/wp-content/uploads/SVM_4-2.png" alt="what image shows"></center>

In such scenarios, we calculate the margin (which is the distance between nearest data point and the hyperplane). The hyperplane with the largest margin will be considered as the correct hyperplane to classify the dataset.

Here C has the largest margin. Hence, it is considered as the best hyperplane.


### Kernels
Knowing 
$$ w = \sum_{k=1}^{N} \alpha_k y_k \varphi(x_k) $$

And
$$ y_{pred} = w^T \varphi(x) + b $$

Then 
$$ y_{pred} = (\sum_{k=1}^{N} \alpha_k y_k \varphi(x_k))^T \varphi(x) + b $$

Where $\varphi$ is a function that makes each input in $x$ correspond to a point in $F$ (a Hilbert space). This can be seen as processing and transforming the input featuers to keep the model's convexity. [2]

This also allows us to transform the inputs into another space where they might be more easily classified.

<center><img src=https://miro.medium.com/max/838/1*gXvhD4IomaC9Jb37tzDUVg.png alt="what image shows"></center>

## ROC and AUC

A ROC (Receiver Operating Characteristic) is a graph that shows how the classification model performs at the classification thresholds. 

ROC curves typically feature true positive rate on the Y axis, and false positive rate on the X axis. This means that the top left corner of the plot is the “ideal” point - a false positive rate of zero, and a true positive rate of one. This is not very realistic, but it does mean that a larger area under the curve (AUC) is usually better. [3]

True Positive Rate is a synonym for Recall and defined as:
$$ TPR = \frac{TP}{TP + FN} $$

False Positive Rate is a synonym for Specificity and defined as:

$$ FPR = \frac{FP}{FP + TN} $$

ROC curves are typically used in binary classification to study the output of a classifier. In order to extend ROC curve and ROC area to multi-label classification, it is necessary to binarize the output. One ROC curve can be drawn per label, but one can also draw a ROC curve by considering each element of the label indicator matrix as a binary prediction (micro-averaging).

E.g. If you lower a classification threshold, more items would be classified as positive, increasing False Positives and True Positives.

AUC stands for Area under the ROC.

## Ejercicio 1

- Utiliza el dataset `Iris`, modela con SVC y haz Cross-Validation de diferentes kernels ('linear', 'poly', 'rbf', 'sigmoid').
- Modela con LogisticRegression.
- El método de Cross-Validation es K-Folds con $k=10$.
- Utiliza el AUC como métrico de Cross-Validation.
- Compara resultados.

In [2]:
from sklearn.datasets import load_iris
from sklearn.model_selection import cross_val_score, KFold
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
import numpy as np

In [3]:
df = load_iris()
X, y = df.data, df.target

In [4]:
kf = KFold(n_splits=10, shuffle=True, random_state=137)

In [5]:
kernels = ['linear', 'poly', 'rbf', 'sigmoid']
resultados_svc = {}

In [6]:
for kernel in kernels:
    modelo = SVC(kernel=kernel, probability=True, random_state=137)
    scores = cross_val_score(modelo, X, y, cv=kf, scoring='roc_auc_ovr')  
    resultados_svc[kernel] = np.mean(scores)

In [7]:
modelo_lr = LogisticRegression(max_iter=200, random_state=137)
scores_lr = cross_val_score(modelo_lr, X, y, cv=kf, scoring='roc_auc_ovr')
resultado_lr = np.mean(scores_lr)

In [8]:
print("Resultados de Cross-Validation")
for kernel, auc in resultados_svc.items():
    print(f"SVC ({kernel}): {auc:.4f}")
print(f"Logistic Regression: {resultado_lr:.4f}")


Resultados de Cross-Validation
SVC (linear): 0.9974
SVC (poly): 0.9974
SVC (rbf): 0.9974
SVC (sigmoid): 0.7363
Logistic Regression: 0.9974


Todos tuvieron los mismos resultados, el unico que no fue el de sigmoid, entonces se podria decir que se puede escoger cualquier modelo menos el de sigmoid

## Ejercicio 2
- Repite el ejercicio 1 con el dataset `Default`. Utiliza `default` como target.

In [9]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder, StandardScaler

In [10]:
data = pd.read_csv("Default.csv")
data.head()

default student      balance        income
0      No      No   729.526495  44361.625074
1      No     Yes   817.180407  12106.134700
2      No      No  1073.549164  31767.138950
3      No      No   529.250605  35704.493940
4      No      No   785.655883  38463.495880

In [11]:
data['default'] = LabelEncoder().fit_transform(data['default']) 
data['student'] = LabelEncoder().fit_transform(data['student']) 

In [12]:
data.head()

default  student      balance        income
0        0        0   729.526495  44361.625074
1        0        1   817.180407  12106.134700
2        0        0  1073.549164  31767.138950
3        0        0   529.250605  35704.493940
4        0        0   785.655883  38463.495880

In [13]:
X = data[['student', 'balance', 'income']]
y = data['default']

In [14]:
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [15]:
kf = KFold(n_splits=10, shuffle=True, random_state=137)

In [16]:
kernels = ['linear', 'poly', 'rbf', 'sigmoid']
resultados_svc = {}

In [17]:
for kernel in kernels:
    modelo = SVC(kernel=kernel, probability=True, random_state=137)
    scores = cross_val_score(modelo, X_scaled, y, cv=kf, scoring='roc_auc')
    resultados_svc[kernel] = np.mean(scores)

In [18]:
modelo_lr = LogisticRegression(max_iter=200, random_state=42)
scores_lr = cross_val_score(modelo_lr, X_scaled, y, cv=kf, scoring='roc_auc')
resultado_lr = np.mean(scores_lr)

In [19]:
print("Resultados de Cross-Validation")
for kernel, auc in resultados_svc.items():
    print(f"SVC ({kernel}): {auc:.4f}")
print(f"Logistic Regression: {resultado_lr:.4f}")

Resultados de Cross-Validation
SVC (linear): 0.9233
SVC (poly): 0.8789
SVC (rbf): 0.8409
SVC (sigmoid): 0.7401
Logistic Regression: 0.9493


Ahora si cambio, el peor sigue siendo el de sigmoid, y en este caso el la mejor opcion de los kernel para SVC fue el linear con .9233. Aunque el de la regresión logística superó a todos con yn .9493

# Addendum

Métricos disponibles para clasificación:
- ‘accuracy’
- ‘balanced_accuracy’
- ‘top_k_accuracy’
- ‘average_precision’
- ‘neg_brier_score’
- ‘f1’
- ‘f1_micro’
- ‘f1_macro’
- ‘f1_weighted’
- ‘f1_samples’
- ‘neg_log_loss’
- ‘precision’ etc.
- ‘recall’ etc.
- ‘jaccard’ etc.
- ‘roc_auc’
- ‘roc_auc_ovr’
- ‘roc_auc_ovo’
- ‘roc_auc_ovr_weighted’
- ‘roc_auc_ovo_weighted’
- ‘d2_log_loss_score’

# References

[1] Shigeo Abe.Support Vector Machines for Pattern Classification,2Ed.Springer-Verlag London,2010. ISBN978-1-84996-097-7. URLhttps://www.springer.com/gp/book/9781849960977.

[2] Johan A K Suykens, Tony Van Gestel, Jos De Brabanter, BartDe Moor, and Joos Vandewalle.Least Squares Support VectorMachines. World Scientific,2002. ISBN9789812381514. URLhttps://www.worldscientific.com/worldscibooks/10.1142/5089.

[3] Bradley, A. P. (1997). The use of the area under the ROC curve in the evaluation of machine learning algorithms. Pattern recognition, 30(7), 1145-1159. URL https://www.researchgate.net/post/how_can_I_interpret_the_ROC_curve_result